tracer/src/Datadog.Trace/Iast/ReturnedHeadersAnalyzer.cs

```diff
@@ -39,6 +41,140 @@ internal static void Analyze(IHeaderDictionary responseHeaders, IntegrationId in
         AnalyzeUnvalidatedRedirect(responseHeaders, integrationId);
+        AnalyzeHeaderInjectionVulnerability(responseHeaders, integrationId);
+    }
+
+    // In header injections, we should exclude some headers to prevent false positives:
+    // location: it is already reported in UNVALIDATED_REDIRECT vulnerability detection.
+    // Sec-WebSocket-Location, Sec-WebSocket-Accept, Upgrade, Connection: Usually the framework gets info from request
+    // access-control-allow-*: when the source of the tainted range is the request header origin or access-control-request-headers
+    // set-cookie: We should ignore set-cookie header if the source of all the tainted ranges are cookies
+    // We should exclude the injection when the tainted string only has one range which comes from a request header with the same name that the header that we are checking in the response.
+    // Headers could store sensitive information, we should redact whole <header_value> if:
+    // <header_name> matches with a RegExp
+    // <header_value> matches with  a RegExp
+    // We should redact the sensitive information from the evidence when:
+    // Tainted range is considered sensitive value
+
+#if NETFRAMEWORK
+    private static void AnalyzeHeaderInjectionVulnerability(NameValueCollection responseHeaders, IntegrationId integrationId)
+#else
+    private static void AnalyzeHeaderInjectionVulnerability(IHeaderDictionary responseHeaders, IntegrationId integrationId)
+#endif
+    {
+        try
+        {
+            IastModule.OnExecutedSinkTelemetry(IastInstrumentedSinks.HeaderInjection);
+
+            var tracer = Tracer.Instance;
+            var currentSpan = (tracer.ActiveScope as Scope)?.Span;
+            var traceContext = currentSpan?.Context?.TraceContext;
+
+#if NETFRAMEWORK
+            foreach (var headerKey in responseHeaders.AllKeys)
+#else
+            foreach (var headerKey in responseHeaders.Keys)
+#endif
+            {
+                if (string.IsNullOrWhiteSpace(headerKey))
+                {
+                    continue;
+                }
+
+                bool isHeaderInjectionException = false;
+                foreach (var excludeType in headerInjectionExceptions)
+                {
+                    if (excludeType.Equals((string)headerKey, StringComparison.OrdinalIgnoreCase))
+                    {
+                        isHeaderInjectionException = true;
+                        break;
+                    }
+                }
+
+                if (isHeaderInjectionException)
+                {
+                    continue;
+                }
+
+                string headerValue = responseHeaders[headerKey];
+
+                if (string.IsNullOrWhiteSpace(headerValue))
+                {
+                    continue;
+                }
+
+                var taintedValue = traceContext?.IastRequestContext?.GetTainted(headerValue);
```

Comment:
If `traceContext?.IastRequestContext` is `null`, it will _always_ be null, and this whole loop will never return anything? Maybe it makes sense to move that check to the start of the method etc, and exit if it's not applicable?

Reply:
If traceContext is not null, iastrequestcontext should never be null but still it's better to check it at first. I have added the check at the beginning of the method. Thanks!